src/app/modules/form/admin-form/__tests__/admin-form.service.spec.ts

```diff
@@ -1625,4 +1626,167 @@ describe('admin-form.service', () => {
       )
     })
   })
+
+  describe('updateFormLogic', () => {
+    const logicId1 = new ObjectId()
+    const logicId2 = new ObjectId()
+    const mockEmailFormId = new ObjectId()
+    const mockEncryptFormId = new ObjectId()
+
+    const mockFormLogicOld = {
+      form_logics: [
+        {
+          _id: logicId1,
+          logicType: 'showFields',
+        } as ILogicSchema,
+        {
+          _id: logicId2,
+          logicType: 'showFields',
+        } as ILogicSchema,
+      ],
+    }
+
+    const updatedLogic = {
+      _id: logicId1,
+      logicType: 'preventSubmit',
+    } as ILogicSchema
+
+    const mockFormLogicUpdated = {
+      form_logics: [
+        {
+          _id: logicId1,
+          logicType: 'preventSubmit',
+        } as ILogicSchema,
+        {
+          _id: logicId2,
+          logicType: 'showFields',
+        } as ILogicSchema,
+      ],
+    }
+
+    const UPDATE_SPY = jest.spyOn(FormModel, 'updateFormLogic')
+
+    let mockEmailForm: IPopulatedForm,
+      mockEncryptForm: IPopulatedForm,
+      mockEmailFormUpdated: IPopulatedForm,
+      mockEncryptFormUpdated: IPopulatedForm
+
+    beforeEach(() => {
+      mockEmailForm = ({
+        _id: mockEmailFormId,
+        status: Status.Public,
+        responseMode: ResponseMode.Email,
+        ...mockFormLogicOld,
+      } as unknown) as IPopulatedForm
+      mockEncryptForm = ({
+        _id: mockEncryptFormId,
+        status: Status.Public,
+        responseMode: ResponseMode.Encrypt,
+        ...mockFormLogicOld,
+      } as unknown) as IPopulatedForm
+      mockEmailFormUpdated = ({
+        ...mockEmailForm,
+        ...mockFormLogicUpdated,
+      } as unknown) as IPopulatedForm
+      mockEncryptFormUpdated = ({
+        ...mockEncryptForm,
+        ...mockFormLogicUpdated,
+      } as unknown) as IPopulatedForm
+    })
+
+    it('should return ok(updated logic) on successful form logic update for email mode form', async () => {
+      // Arrange
+      const STATIC_UPDATE_SPY = jest
+        .spyOn(FormModel, 'findByIdAndUpdate')
```

Comment:
is it easier to write the tests if you spy on `updateFormLogic` instead?

Reply:
thanks, its already spied on in L1667 and checked in L1730, etc